server/src/main/java/io/druid/metadata/IndexerSQLMetadataStorageCoordinator.java

```diff
@@ -359,9 +362,10 @@ public SegmentPublishResult inTransaction(
       );
     }
     catch (CallbackFailedException e) {
-      if (txnFailure.get()) {
-        return new SegmentPublishResult(ImmutableSet.of(), false);
+      if (definitelyNotUpdated.get()) {
+        return SegmentPublishResult.fail();
```

Comment:
What do you think about adding an exception to `SegmentPublishResult` on failure, so that callers can figure out why it failed?

Reply:
I think it's not necessary, since there is supposed to be only one reason: compare-and-swap failure with the metadata update.